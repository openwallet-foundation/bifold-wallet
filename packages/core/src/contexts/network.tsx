import { NetInfoStateType, useNetInfo } from '@react-native-community/netinfo'
import { createContext, useContext, useState, useCallback, PropsWithChildren } from 'react'

import NetInfoModal from '../components/modals/NetInfoModal'

export interface NetworkContext {
  silentAssertConnectedNetwork: () => boolean | null
  assertNetworkConnected: () => boolean
  displayNetInfoModal: () => void
  hideNetInfoModal: () => void
  assertInternetReachable: () => boolean | null
}

export const NetworkContext = createContext<NetworkContext>(null as unknown as NetworkContext)

// NOTE: @react-native-community/netinfo can be configured to use whichever reachability check desired
// eg. isInternetReachable can be set to check a specific URL (like your mediator). See the docs here for more info:
// https://github.com/react-native-netinfo/react-native-netinfo?tab=readme-ov-file#configure
export const NetworkProvider = ({ children }: PropsWithChildren) => {
  const { isConnected, type, isInternetReachable } = useNetInfo()
  const [isNetInfoModalDisplayed, setIsNetInfoModalDisplayed] = useState<boolean>(false)

  const displayNetInfoModal = useCallback(() => {
    setIsNetInfoModalDisplayed(true)
  }, [])

  const hideNetInfoModal = useCallback(() => {
    setIsNetInfoModalDisplayed(false)
  }, [])

  /**
   * Returns null until the network state is known, then returns boolean
   * Useful for cases where we do not want to take action until the network state is known
   *
   * @returns {boolean | null} - `true` if the network is connected, `false` if not connected,
   *                             and `null` if the network status not yet known
   */
  const silentAssertConnectedNetwork = useCallback((): boolean | null => {
    return type === NetInfoStateType.unknown
      ? null
      : isConnected || [NetInfoStateType.wifi, NetInfoStateType.cellular].includes(type)
  }, [isConnected, type])

  /**
   * Strictly asserts that the network is connected. Will return false even if
   * the network state is not yet known - in this case it will also display the
   * NetInfoModal
   * Useful for cases where we must be sure of connectivity before proceeding
   *
   * @returns {boolean} - `true` if the network is checked and connected, otherwise `false`
   */
  const assertNetworkConnected = useCallback(() => {
    const connectionConfirmed = silentAssertConnectedNetwork() === true
    if (!connectionConfirmed) {
      displayNetInfoModal()
    }

    return connectionConfirmed
  }, [silentAssertConnectedNetwork, displayNetInfoModal])

  const assertInternetReachable = useCallback((): boolean | null => {
    return isInternetReachable
  }, [isInternetReachable])

  return (
    <NetworkContext.Provider
      value={{
        silentAssertConnectedNetwork,
        assertNetworkConnected,
        displayNetInfoModal,
        hideNetInfoModal,
        assertInternetReachable,
      }}
    >
      {children}
      <NetInfoModal visible={isNetInfoModalDisplayed} onSubmit={() => hideNetInfoModal()} />
    </NetworkContext.Provider>
  )
}

export const useNetwork = () => useContext(NetworkContext)
