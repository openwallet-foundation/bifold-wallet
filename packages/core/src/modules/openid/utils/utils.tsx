import { getDomainFromUrl } from '@credo-ts/core'
import { Attribute, Field } from '@hyperledger/aries-oca/build/legacy'
import { W3cCredentialDisplay } from '../types'

/**
 * Converts a camelCase string to a sentence format (first letter capitalized, rest in lower case).
 * i.e. sanitizeString("helloWorld")  // returns: 'Hello world'
 */
export function sanitizeString(str: string) {
  const result = str.replace(/([a-z0-9])([A-Z])/g, '$1 $2').replaceAll('_', ' ')
  let words = result.split(' ')
  words = words.map((word, index) => {
    if (index === 0 || word.toUpperCase() === word) {
      return word.charAt(0).toUpperCase() + word.slice(1)
    }
    return word.charAt(0).toLowerCase() + word.slice(1)
  })
  return words.join(' ')
}

export function getHostNameFromUrl(url: string) {
  try {
    return getDomainFromUrl(url)
  } catch (error) {
    throw new Error(`Error getting hostname from url: ${error}`)
  }
}

export const buildFieldsFromOpenIDTemplate = (data: { [key: string]: unknown }): Array<Field> => {
  const fields = []
  for (const key of Object.keys(data)) {
    // omit id and type
    if (key === 'id' || key === 'type') continue

    let pushedVal: string | number | null = null
    if (typeof data[key] === 'string' || typeof data[key] === 'number') {
      pushedVal = data[key] as string | number | null
    }
    fields.push(new Attribute({ name: key, value: pushedVal }))
  }
  return fields
}

export function formatDate(input: string | Date): string {
  const date = input instanceof Date ? input : new Date(input)
  return date.toLocaleString('en-US', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: true,
  })
}

export const getAttributeField = (display: W3cCredentialDisplay, searchKey: string): Attribute | undefined => {
  for (const [key, value] of Object.entries(display.attributes)) {
    if (searchKey === key) {
      return new Attribute({
        name: key,
        value: value as string | number | null,
        mimeType: typeof value === 'number' ? 'text/number' : 'text/plain',
      })
    }
  }
  return undefined
}

/**
 * Mostly used for mdoc crednetials
 */
export function detectImageMimeType(data: Uint8Array): 'image/jpeg' | 'image/jp2' | null {
  // Check if array is too short to contain magic numbers
  if (data.length < 12) {
    return null
  }

  // JPEG starts with FF D8 FF
  if (data[0] === 0xff && data[1] === 0xd8 && data[2] === 0xff) {
    return 'image/jpeg'
  }

  // JPEG2000 has two possible signatures:
  // 1) 00 00 00 0C 6A 50 20 20 0D 0A 87 0A
  // 2) FF 4F FF 51

  // Check first signature
  if (
    data[0] === 0x00 &&
    data[1] === 0x00 &&
    data[2] === 0x00 &&
    data[3] === 0x0c &&
    data[4] === 0x6a && // 'j'
    data[5] === 0x50 && // 'P'
    data[6] === 0x20 &&
    data[7] === 0x20 &&
    data[8] === 0x0d &&
    data[9] === 0x0a &&
    data[10] === 0x87 &&
    data[11] === 0x0a
  ) {
    return 'image/jp2'
  }

  // Check second signature
  if (data[0] === 0xff && data[1] === 0x4f && data[2] === 0xff && data[3] === 0x51) {
    return 'image/jp2'
  }

  return null
}

/**
 * very simple matcher for `yyyy-mm-dd`
 */
export function isDateString(value: string) {
  // We do the length check first to avoid unnecesary regex
  return value.length === 'yyyy-mm-dd'.length && value.match(/^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$/)
}

export function isW3CProofRequest(type: string): boolean {
  if (type === 'jwt_vc') {
    return true
  }
  return false
}

export function isSdJwtProofRequest(type: string): boolean {
  if (type === 'vc+sd-jwt') {
    return true
  }
  return false
}
