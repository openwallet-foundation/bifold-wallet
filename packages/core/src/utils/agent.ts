import {
  AnonCredsCredentialFormatService,
  AnonCredsModule,
  AnonCredsProofFormatService,
  DataIntegrityCredentialFormatService,
  LegacyIndyCredentialFormatService,
  LegacyIndyProofFormatService,
  V1CredentialProtocol,
  V1ProofProtocol,
} from '@credo-ts/anoncreds'
import { AskarModule } from '@credo-ts/askar'
import {
  Agent,
  AutoAcceptCredential,
  AutoAcceptProof,
  ConnectionsModule,
  CredentialsModule,
  DifPresentationExchangeProofFormatService,
  MediationRecipientModule,
  MediatorPickupStrategy,
  ProofsModule,
  V2CredentialProtocol,
  V2ProofProtocol,
} from '@credo-ts/core'
import { IndyVdrAnonCredsRegistry, IndyVdrModule, IndyVdrPoolConfig } from '@credo-ts/indy-vdr'
import { OpenId4VcHolderModule } from '@credo-ts/openid4vc'
import { PushNotificationsApnsModule, PushNotificationsFcmModule } from '@credo-ts/push-notifications'
import { useAgent } from '@credo-ts/react-hooks'
import { anoncreds } from '@hyperledger/anoncreds-react-native'
import { ariesAskar } from '@hyperledger/aries-askar-react-native'
import { indyVdr } from '@hyperledger/indy-vdr-react-native'

interface GetAgentModulesOptions {
  indyNetworks: IndyVdrPoolConfig[]
  mediatorInvitationUrl?: string
  txnCache?: { capacity: number; expiryOffsetMs: number; path?: string }
}

export type BifoldAgent = Agent<ReturnType<typeof getAgentModules>>

/**
 * Constructs the modules to be used in the agent setup
 * @param indyNetworks
 * @param mediatorInvitationUrl determine which mediator to use
 * @param txnCache optional local cache config for indyvdr
 * @returns modules to be used in agent setup
 */
export function getAgentModules({ indyNetworks, mediatorInvitationUrl, txnCache }: GetAgentModulesOptions) {
  const indyCredentialFormat = new LegacyIndyCredentialFormatService()
  const indyProofFormat = new LegacyIndyProofFormatService()

  if (txnCache) {
    indyVdr.setLedgerTxnCache({
      capacity: txnCache.capacity,
      expiry_offset_ms: txnCache.expiryOffsetMs,
      path: txnCache.path,
    })
  }

  return {
    askar: new AskarModule({
      ariesAskar,
    }),
    anoncreds: new AnonCredsModule({
      anoncreds,
      registries: [new IndyVdrAnonCredsRegistry()],
    }),
    indyVdr: new IndyVdrModule({
      indyVdr,
      networks: indyNetworks as [IndyVdrPoolConfig],
    }),
    connections: new ConnectionsModule({
      autoAcceptConnections: true,
    }),
    credentials: new CredentialsModule({
      autoAcceptCredentials: AutoAcceptCredential.ContentApproved,
      credentialProtocols: [
        new V1CredentialProtocol({ indyCredentialFormat }),
        new V2CredentialProtocol({
          credentialFormats: [
            indyCredentialFormat,
            new AnonCredsCredentialFormatService(),
            new DataIntegrityCredentialFormatService(),
          ],
        }),
      ],
    }),
    proofs: new ProofsModule({
      autoAcceptProofs: AutoAcceptProof.ContentApproved,
      proofProtocols: [
        new V1ProofProtocol({ indyProofFormat }),
        new V2ProofProtocol({
          proofFormats: [
            indyProofFormat,
            new AnonCredsProofFormatService(),
            new DifPresentationExchangeProofFormatService(),
          ],
        }),
      ],
    }),
    mediationRecipient: new MediationRecipientModule({
      mediatorInvitationUrl: mediatorInvitationUrl,
      mediatorPickupStrategy: MediatorPickupStrategy.Implicit,
    }),
    pushNotificationsFcm: new PushNotificationsFcmModule(),
    pushNotificationsApns: new PushNotificationsApnsModule(),
    openId4VcHolder: new OpenId4VcHolderModule(),
  }
}

interface MyAgentContextInterface {
  loading: boolean
  agent?: BifoldAgent
  setAgent: (agent?: BifoldAgent) => void
}

export const useAppAgent = useAgent as () => MyAgentContextInterface

export const createLinkSecretIfRequired = async (agent: Agent) => {
  // If we don't have any link secrets yet, we will create a
  // default link secret that will be used for all anoncreds
  // credential requests.
  const linkSecretIds = await agent.modules.anoncreds.getLinkSecretIds()
  if (linkSecretIds.length === 0) {
    await agent.modules.anoncreds.createLinkSecret({
      setAsDefault: true,
    })
  }
}
