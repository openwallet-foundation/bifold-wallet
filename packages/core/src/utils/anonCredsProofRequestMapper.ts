import {
  AnonCredsCredentialTags,
  AnonCredsCredentialsForProofRequest,
  AnonCredsProofRequest,
  AnonCredsProofRequestRestriction,
  AnonCredsRequestedAttribute,
  AnonCredsRequestedPredicate,
  getAnonCredsTagsFromRecord,
} from '@credo-ts/anoncreds'
import {
  ClaimFormat,
  DifPexCredentialsForRequest,
  DifPresentationExchangeDefinition,
  DifPresentationExchangeDefinitionV2,
  W3cCredentialRecord,
} from '@credo-ts/core'

import { ProofCredentialAttributes, ProofCredentialPredicates } from '../types/proof-items'

export type RecordWithMetadata = {
  record: W3cCredentialRecord
  anonCredsTags: AnonCredsCredentialTags
}
export type DescriptorMetadata = { [key: string]: RecordWithMetadata[] }

export interface DifPexAnonCredsProofRequest extends AnonCredsProofRequest {
  requested_attributes: Record<string, AnonCredsRequestedAttribute & { descriptorId?: string }>
  requested_predicates: Record<string, AnonCredsRequestedPredicate & { descriptorId?: string }>
}

type FieldV2 = NonNullable<
  DifPresentationExchangeDefinitionV2['input_descriptors'][number]['constraints']['fields']
>[number]

const getPredicateTypeAndValues = (predicateFilter: NonNullable<FieldV2['filter']>) => {
  const predicates: {
    predicateType: AnonCredsRequestedPredicate['p_type']
    predicateValue: AnonCredsRequestedPredicate['p_value']
  }[] = []

  const supportedJsonSchemaNumericRangeProperties: Record<string, AnonCredsRequestedPredicate['p_type']> = {
    exclusiveMinimum: '>',
    exclusiveMaximum: '<',
    minimum: '>=',
    maximum: '<=',
  }

  for (const [key, value] of Object.entries(predicateFilter)) {
    if (key === 'type') continue

    const predicateType = supportedJsonSchemaNumericRangeProperties[key]
    if (!predicateType) throw new Error(`Unsupported predicate filter property '${key}'`)
    predicates.push({ predicateType, predicateValue: value })
  }

  return predicates
}

const getClaimNameForField = (field: FieldV2) => {
  if (!field.path) throw new Error('Field path is required')
  const baseClaimPath = '$.credentialSubject.'
  const claimPaths = field.path.filter((path) => path.startsWith(baseClaimPath))
  if (claimPaths.length === 0) return undefined

  const claimNames = claimPaths.map((path) => path.slice(baseClaimPath.length))
  const propertyName = claimNames[0]

  return propertyName
}

export type ProcessedAttributes = { [key: string]: ProofCredentialAttributes }
export type ProcessedPredicates = { [key: string]: ProofCredentialPredicates }

export const createAnonCredsProofRequest = (
  presentationDefinition: DifPresentationExchangeDefinition,
  descriptorMetadata: DescriptorMetadata
) => {
  const anonCredsProofRequest: DifPexAnonCredsProofRequest = {
    version: '1.0',
    name: presentationDefinition.name ?? 'Proof request',
    nonce: 'nonce',
    requested_attributes: {},
    requested_predicates: {},
  }

  const nonRevokedTime = Math.floor(Date.now() / 1000)
  const nonRevokedInterval = { from: nonRevokedTime, to: nonRevokedTime }

  for (const descriptor of presentationDefinition.input_descriptors) {
    const referent = descriptor.id
    let predicateReferentIndex = 0

    const fields = descriptor.constraints?.fields
    if (!fields) throw new Error('Unclear mapping of constraint with no fields.')

    // Setting `requiresRevocationStatus` to `false` returns all
    // credentials even if they are revokable (and revoked). We need this to
    // be able to show why a proof cannot be satisfied. Otherwise we can only
    // show failure.
    const requiresRevocationStatus = false

    const credentialMeta = descriptorMetadata[descriptor.id]
    const restrictions: AnonCredsProofRequestRestriction[] = credentialMeta.map((credentialMeta) => {
      return {
        cred_def_id: credentialMeta.anonCredsTags.anonCredsCredentialDefinitionId,
        schema_id: credentialMeta.anonCredsTags.anonCredsSchemaId,
      }
    })

    for (const field of fields) {
      const propertyName = getClaimNameForField(field)
      if (!propertyName) continue

      if (field.predicate) {
        if (!field.filter) throw new Error('Missing required predicate filter property.')
        const predicateTypeAndValues = getPredicateTypeAndValues(field.filter)
        for (const { predicateType, predicateValue } of predicateTypeAndValues) {
          const predicateReferent = `${referent}_${predicateReferentIndex++}`
          anonCredsProofRequest.requested_predicates[predicateReferent] = {
            name: propertyName,
            p_type: predicateType,
            p_value: predicateValue,
            restrictions,
            non_revoked: requiresRevocationStatus ? nonRevokedInterval : undefined,
            descriptorId: descriptor.id,
          }
        }
      } else {
        if (!anonCredsProofRequest.requested_attributes[referent]) {
          anonCredsProofRequest.requested_attributes[referent] = {
            names: [propertyName],
            restrictions,
            non_revoked: requiresRevocationStatus ? nonRevokedInterval : undefined,
            descriptorId: descriptor.id,
          }
        } else {
          const names = anonCredsProofRequest.requested_attributes[referent].names ?? []
          anonCredsProofRequest.requested_attributes[referent].names = [...names, propertyName]
        }
      }
    }
  }

  return anonCredsProofRequest
}

export const getDescriptorMetadata = (credentialsForRequest: DifPexCredentialsForRequest) => {
  const descriptorMetadata: DescriptorMetadata = {}

  for (const requestedAttribute of Object.values(credentialsForRequest.requirements)) {
    for (const entry of requestedAttribute.submissionEntry) {
      const inputDescriptorId = entry.inputDescriptorId

      const recordsWithMetadata = entry.verifiableCredentials.map((submissionEntryCredential) => {
        if (submissionEntryCredential.type !== ClaimFormat.LdpVc) {
          throw new Error(`Unsupported credential type. ${submissionEntryCredential.type}`)
        }
        const record = submissionEntryCredential.credentialRecord
        const anonCredsTags = getAnonCredsTagsFromRecord(record as W3cCredentialRecord)
        if (!anonCredsTags) throw new Error('Missing AnonCreds tags from credential record')
        return { record, anonCredsTags }
      })

      if (!descriptorMetadata[inputDescriptorId]) descriptorMetadata[inputDescriptorId] = []

      for (const recordWithMetadata of recordsWithMetadata) {
        if (!descriptorMetadata[inputDescriptorId].find((item) => item.record.id === recordWithMetadata.record.id)) {
          descriptorMetadata[inputDescriptorId].push(recordWithMetadata)
        }
      }
    }
  }

  return descriptorMetadata
}

/**
 * The matches returned by our artificial anonCredsProofRequest could contain matches,
 * which are not valid thus we need to filter them out
 */
export const filterInvalidProofRequestMatches = (
  anonCredsCredentialsForRequest: AnonCredsCredentialsForProofRequest,
  descriptorMetadata: DescriptorMetadata
) => {
  anonCredsCredentialsForRequest.attributes = Object.fromEntries(
    Object.entries(anonCredsCredentialsForRequest.attributes).map(([referent, matches]) => {
      const descriptorMeta = descriptorMetadata[referent]
      const validCredentialsForReferent = descriptorMeta.map((meta) => meta.record.id)
      const validMatches = matches.filter((match) => validCredentialsForReferent.includes(match.credentialId))
      return [referent, validMatches]
    })
  )

  anonCredsCredentialsForRequest.predicates = Object.fromEntries(
    Object.entries(anonCredsCredentialsForRequest.predicates).map(([_referent, matches]) => {
      const referent = _referent.split('_').slice(0, -1).join('_')
      const descriptorMeta = descriptorMetadata[referent]
      const validCredentialsForReferent = descriptorMeta.map((meta) => meta.record.id)
      const validMatches = matches.filter((match) => validCredentialsForReferent.includes(match.credentialId))
      return [_referent, validMatches]
    })
  )

  return anonCredsCredentialsForRequest
}
