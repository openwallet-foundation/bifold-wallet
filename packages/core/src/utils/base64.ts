// React Native compatible base64 encoding/decoding
// Uses global btoa/atob if available, otherwise falls back to polyfill

export function toBase64(bytes: Uint8Array): string {
  let binary = ''
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i])
  }
  if (typeof btoa !== 'undefined') {
    return btoa(binary)
  }
  // Polyfill
  return globalThis.btoa ? globalThis.btoa(binary) : base64EncodePolyfill(binary)
}

export function fromBase64(base64: string): Uint8Array {
  let binary = ''
  if (typeof atob !== 'undefined') {
    binary = atob(base64)
  } else if (globalThis.atob) {
    binary = globalThis.atob(base64)
  } else {
    binary = base64DecodePolyfill(base64)
  }
  const bytes = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i)
  }
  return bytes
}

// Minimal base64 polyfill for fallback
function base64EncodePolyfill(str: string): string {
  // This is a minimal, non-URL-safe polyfill for base64 encoding
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/='
  let output = ''
  let i = 0
  while (i < str.length) {
    const chr1 = str.charCodeAt(i++)
    const chr2 = str.charCodeAt(i++)
    const chr3 = str.charCodeAt(i++)
    const enc1 = chr1 >> 2
    const enc2 = ((chr1 & 3) << 4) | (chr2 >> 4)
    const enc3 = isNaN(chr2) ? 64 : ((chr2 & 15) << 2) | (chr3 >> 6)
    const enc4 = isNaN(chr3) ? 64 : chr3 & 63
    output += chars.charAt(enc1) + chars.charAt(enc2) + chars.charAt(enc3) + chars.charAt(enc4)
  }
  return output
}

function base64DecodePolyfill(input: string): string {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/='
  const str = input.replace(/=+$/, '')
  let output = ''
  let i = 0
  while (i < str.length) {
    const enc1 = chars.indexOf(str.charAt(i++))
    const enc2 = chars.indexOf(str.charAt(i++))
    const enc3 = chars.indexOf(str.charAt(i++))
    const enc4 = chars.indexOf(str.charAt(i++))
    const chr1 = (enc1 << 2) | (enc2 >> 4)
    const chr2 = ((enc2 & 15) << 4) | (enc3 >> 2)
    const chr3 = ((enc3 & 3) << 6) | enc4
    output += String.fromCharCode(chr1)
    if (enc3 !== 64) output += String.fromCharCode(chr2)
    if (enc4 !== 64) output += String.fromCharCode(chr3)
  }
  return output
}
