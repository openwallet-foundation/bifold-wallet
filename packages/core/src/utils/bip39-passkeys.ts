/**
 * BIP39 Deterministic Passkey Integration for Rocca Wallet
 *
 * This utility integrates the hashmap-credo-core BIP39 wallet functionality
 * to generate deterministic P-256 passkeys for WebAuthn authentication.
 */

import {
  createDeterministicP256Key,
  enhancedCreateKey,
  type DeterministicP256Options,
  type BIP39WalletCreateKeyOptions,
} from '@credo-ts/core'
import { WalletApi } from '@credo-ts/core'
import type { BifoldAgent } from './agent'

/**
 * Options for configuring deterministic passkey generation
 */
export interface PasskeyGenerationOptions {
  /** BIP39 mnemonic phrase for deterministic generation */
  mnemonic: string
  /** WebAuthn origin/domain (e.g., "https://example.com") */
  origin: string
  /** WebAuthn user handle/identifier */
  userHandle: string
  /** Optional counter for generating multiple keys */
  counter?: number
}

/**
 * Enhanced wallet API that supports deterministic P-256 passkey generation
 */
export class BIP39PasskeyWallet {
  private originalWallet: WalletApi
  private defaultPasskeyOptions?: PasskeyGenerationOptions

  constructor(originalWallet: WalletApi, defaultOptions?: PasskeyGenerationOptions) {
    this.originalWallet = originalWallet
    this.defaultPasskeyOptions = defaultOptions
  }

  /**
   * Create a key with optional deterministic P-256 passkey generation
   */
  async createKey(options: BIP39WalletCreateKeyOptions): Promise<import('@credo-ts/core').Key> {
    return enhancedCreateKey(this.originalWallet.createKey.bind(this.originalWallet), options)
  }

  /**
   * Generate a deterministic P-256 passkey for WebAuthn
   */
  async createDeterministicPasskey(options: PasskeyGenerationOptions): Promise<import('@credo-ts/core').Key> {
    const p256Options: DeterministicP256Options = {
      mnemonic: options.mnemonic,
      origin: options.origin,
      userHandle: options.userHandle,
      counter: options.counter || 0,
    }

    return createDeterministicP256Key(p256Options)
  }

  /**
   * Create a P-256 passkey using default options if configured
   */
  async createDefaultPasskey(userHandle?: string, counter?: number): Promise<import('@credo-ts/core').Key> {
    if (!this.defaultPasskeyOptions) {
      throw new Error('No default passkey options configured. Use createDeterministicPasskey() with explicit options.')
    }

    return this.createDeterministicPasskey({
      ...this.defaultPasskeyOptions,
      userHandle: userHandle || this.defaultPasskeyOptions.userHandle,
      counter: counter || this.defaultPasskeyOptions.counter || 0,
    })
  }

  // Delegate all other methods to the original wallet
  get isInitialized() {
    return this.originalWallet.isInitialized
  }
  get isProvisioned() {
    return this.originalWallet.isProvisioned
  }
  get walletConfig() {
    return this.originalWallet.walletConfig
  }

  async initialize(options: any) {
    return this.originalWallet.initialize(options)
  }
  async create(options: any) {
    return this.originalWallet.create(options)
  }
  async createAndOpen(options: any) {
    return this.originalWallet.createAndOpen(options)
  }
  async open(options: any) {
    return this.originalWallet.open(options)
  }
  async rotateKey(options: any) {
    return this.originalWallet.rotateKey(options)
  }
  async close() {
    return this.originalWallet.close()
  }
  async delete() {
    return this.originalWallet.delete()
  }
  async export(options: any) {
    return this.originalWallet.export(options)
  }
  async import(walletConfig: any, importConfig: any) {
    return this.originalWallet.import(walletConfig, importConfig)
  }
  async generateNonce() {
    return this.originalWallet.generateNonce()
  }
}

/**
 * Enhance an existing agent with BIP39 deterministic passkey capabilities
 */
export function enhanceAgentWithPasskeys(
  agent: BifoldAgent,
  passkeyOptions?: PasskeyGenerationOptions
): BIP39PasskeyWallet {
  return new BIP39PasskeyWallet(agent.wallet, passkeyOptions)
}

/**
 * Utility to generate a deterministic passkey directly without wrapping the wallet
 */
export async function generateDeterministicPasskey(options: PasskeyGenerationOptions) {
  const p256Options: DeterministicP256Options = {
    mnemonic: options.mnemonic,
    origin: options.origin,
    userHandle: options.userHandle,
    counter: options.counter || 0,
  }

  return createDeterministicP256Key(p256Options)
}

/**
 * Helper to validate BIP39 mnemonic phrases
 */
export function validateMnemonic(mnemonic: string): boolean {
  // Basic validation - should be 12, 15, 18, 21, or 24 words
  const words = mnemonic.trim().split(/\s+/)
  return [12, 15, 18, 21, 24].includes(words.length)
}

/**
 * Generate a deterministic user handle from a base identifier
 * This can be useful for consistent user identification across sessions
 */
export function generateDeterministicUserHandle(baseId: string, domain: string): string {
  // Simple deterministic approach - you might want to use a more sophisticated method
  return `${baseId}@${domain}`
}
