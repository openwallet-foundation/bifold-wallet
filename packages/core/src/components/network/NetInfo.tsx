import { useEffect, useCallback, useState } from 'react'
import { useTranslation } from 'react-i18next'
import Toast from 'react-native-toast-message'
import { useNetwork } from '../../contexts/network'
import { ToastType } from '../toast/BaseToast'

const NetInfo: React.FC = () => {
  const { assertInternetReachable } = useNetwork()
  const { t } = useTranslation()
  const [hasShown, setHasShown] = useState(false)

  const showNetworkWarning = useCallback(() => {
    setHasShown(true)
    Toast.show({
      type: ToastType.Error,
      autoHide: true,
      text1: t('NetInfo.NoInternetConnectionTitle'),
    })
  }, [t])

  useEffect(() => {
    const internetReachable = assertInternetReachable()
    if (internetReachable) {
      Toast.hide()
    }

    // Strict check for false, null means the network state is not yet known
    if (internetReachable === false && !hasShown) {
      showNetworkWarning()
    }
  }, [showNetworkWarning, assertInternetReachable, hasShown])

  return null
}

export default NetInfo
