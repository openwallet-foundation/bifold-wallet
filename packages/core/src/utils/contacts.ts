import { DidCommBasicMessageRecord, DidCommConnectionRecord, DidCommCredentialExchangeRecord, DidCommProofExchangeRecord } from '@credo-ts/didcomm'
import { BifoldAgent } from './agent'

interface ConnectionWithMessages {
  conn: DidCommConnectionRecord
  msgs: (DidCommBasicMessageRecord | DidCommCredentialExchangeRecord | DidCommProofExchangeRecord)[]
}

interface ConnectionWithLatestMessage {
  conn: DidCommConnectionRecord
  latestMsg: DidCommBasicMessageRecord | DidCommCredentialExchangeRecord | DidCommProofExchangeRecord
}

/**
 * Function to fetch contacts (connections) in order of latest chat message without using hooks
 * @param agent - Credo agent
 * @returns ConnectionRecord[] sorted by most recent message
 */
export const fetchContactsByLatestMessage = async (
  agent: BifoldAgent,
  connections: DidCommConnectionRecord[]
): Promise<DidCommConnectionRecord[]> => {
  const connectionsWithMessages = await Promise.all<ConnectionWithMessages>(
    connections.map(
      async (conn: DidCommConnectionRecord): Promise<ConnectionWithMessages> => ({
        conn,
        msgs: [
          ...(await agent.modules.basicMessages.findAllByQuery({ connectionId: conn.id })),
          ...(await agent.modules.proofs.findAllByQuery({ connectionId: conn.id })),
          ...(await agent.modules.credentials.findAllByQuery({ connectionId: conn.id })),
        ],
      })
    )
  )

  const connectionsWithLatestMessage: ConnectionWithLatestMessage[] = connectionsWithMessages.map((pair) => {
    return {
      conn: pair.conn,
      latestMsg: pair.msgs.reduce(
        (acc, cur) => {
          const accDate = acc.updatedAt || acc.createdAt
          const curDate = cur.updatedAt || cur.createdAt
          return accDate > curDate ? acc : cur
        },
        // Initial value if no messages exist for this connection is a placeholder with the date the connection was created
        { createdAt: pair.conn.createdAt } as DidCommBasicMessageRecord | DidCommCredentialExchangeRecord | DidCommProofExchangeRecord
      ),
    }
  })

  return connectionsWithLatestMessage
    .sort(
      (a, b) =>
        new Date(b.latestMsg.updatedAt || b.latestMsg.createdAt).valueOf() -
        new Date(a.latestMsg.updatedAt || a.latestMsg.createdAt).valueOf()
    )
    .map((pair) => pair.conn)
}
