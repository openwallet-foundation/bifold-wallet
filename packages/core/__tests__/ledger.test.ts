import { canConnectToHost, fetchLedgerNodes } from '../src/utils/network'

jest.mock('react-native/Libraries/EventEmitter/NativeEventEmitter')
jest.mock('../src/configs/ledgers/indy')
jest.useFakeTimers({ legacyFakeTimers: true })
jest.spyOn(global, 'setTimeout')

// Use these ports to trigger TCP socket behaviour in the
// mock.
const ports = {
  connect: 8001,
  error: 8002,
  timeout: 8003,
}

describe('Ledger utility', () => {
  test('The default ledger nodes are fetched', async () => {
    const results = fetchLedgerNodes()

    expect(results).toMatchSnapshot()
  })

  test('No nodes are returned for a faux ledger', async () => {
    const results = fetchLedgerNodes('springfield')

    expect(results).toHaveLength(0)
  })

  test('An available host returns "true"', async () => {
    const result = await canConnectToHost({ host: '192.168.100.1', port: ports.connect })

    expect(result).toBe(true)
  })

  test('An un-available host returns "false"', async () => {
    const result = await canConnectToHost({ host: '192.168.100.1', port: ports.timeout })

    expect(result).toBe(false)
  })

  test('An bad host returns "false"', async () => {
    const result = await canConnectToHost({ host: '192.168.100.1', port: ports.error })

    expect(result).toBe(false)
  })
})
