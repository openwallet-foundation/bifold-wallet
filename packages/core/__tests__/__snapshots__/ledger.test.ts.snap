// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ledger utility The default ledger nodes are fetched 1`] = `
Array [
  Object {
    "host": "10.0.0.1",
    "port": 9702,
  },
  Object {
    "host": "192.168.0.0",
    "port": 9702,
  },
  Object {
    "host": "10.0.1.1",
    "port": 9722,
  },
]
`;
